include/pybind11/pytypes.h

```diff
@@ -377,20 +381,54 @@ class PYBIND11_EXPORT_EXCEPTION error_already_set : public std::runtime_error {
 public:
     /// Constructs a new exception from the current Python error indicator, if any.  The current
     /// Python error indicator will be cleared.
-    error_already_set() : std::runtime_error(detail::error_string()) {
+    error_already_set() : std::runtime_error("") {
         PyErr_Fetch(&m_type.ptr(), &m_value.ptr(), &m_trace.ptr());
+        if (m_type) {
+            PyErr_NormalizeException(&m_type.ptr(), &m_value.ptr(), &m_trace.ptr());
+        }
     }
 
-    error_already_set(const error_already_set &) = default;
-    error_already_set(error_already_set &&) = default;
+    error_already_set(const error_already_set &) noexcept = default;
+    error_already_set(error_already_set &&) noexcept = default;
 
     inline ~error_already_set() override;
 
+    const char *what() const noexcept override {
+        auto *sup_what = std::runtime_error::what();
+        if (sup_what[0] != '\0') {
+            return sup_what;
+        } else if (m_lazy_what.empty()) {
+            try {
+                m_lazy_what = detail::error_string(m_type.ptr(), m_value.ptr(), m_trace.ptr());
+            } catch (const std::exception &e) {
+                m_lazy_what
+                    = std::string(
+                          "Unknown internal error occurred while constructing error_string: ")
+                      + e.what();
+                return m_lazy_what.c_str();
+            } catch (...) {
+                m_lazy_what = "Unknown internal error occurred";
+                return m_lazy_what.c_str();
+            }
+        }
+
+        assert(!m_lazy_what.empty());
+        return m_lazy_what.c_str();
+    }
+
     /// Give the currently-held error back to Python, if any.  If there is currently a Python error
     /// already set it is cleared first.  After this call, the current object no longer stores the
     /// error variables (but the `.what()` string is still available).
     void restore() {
-        PyErr_Restore(m_type.release().ptr(), m_value.release().ptr(), m_trace.release().ptr());
+        // Abuse assignment operator to cache what()
+        std::runtime_error::operator=(std::runtime_error(what())); // Force-build `.what()`.
```

Comment:
@henryiii @rwgk It occurred to me we can use the parent's assignment operator to change what() after construction, but using it in a const method requires a const_cast which is undefined behavior. This would make it behave as expected when being up cast after err.restore() is called though.

Reply:
Quoting from https://en.cppreference.com/w/cpp/error/runtime_error

> The pointer is guaranteed to be valid at least until the exception object from which it is obtained is destroyed, or until a non-const member function (e.g. copy assignment operator) on the exception object is called.

From that I think a `const_cast` is compatible with expectations for `std::runtime_error`.

But the upcast idea has me very worried. If what I see depends on the order of the upcasts and `.what()` calls, it could lead to extremely confusing debugging. IOW I'm budgeting wasted human time on that vs the performance gains. The performance gains may never be "actually consequential", nice of course, but the hours lost in confusion, at scale almost a certainty, can really hurt. — I don't think this can be quantified precisely, but I sure know such pain!

I think it would be better to not make this overly smart/sneaky.

Idea: `std::runtime_error(nullptr)` to die if someone tries to use the unavailable`.what()`
Non-starter: `what():  basic_string::_S_construct null not valid`

Next idea (maybe even the better one): `std::runtime_error("This is an up-cast pybind11::error_already_set: details are unavailable. Please catch & inspect the pybind11::error_already_set object.")`

Never change it.

Reply:
WAIT: https://en.cppreference.com/w/cpp/error/exception

`what()` is virtual!

I was looking for that before but misunderstood something.

Reply:
My previous worries are apparently unfounded. See below.
But I must still be missing something: why is the simple `what()` at the beginning of `restore()` not sufficient?

```
$ cat error_already_set_up_cast.cpp 
#include <iostream>
#include <stdexcept>
#include <string>

struct error_already_set : std::runtime_error {
    error_already_set() : std::runtime_error("Do not look here.") {}
    const char *what() const noexcept override {
        num_what_calls++;
        curr_what = std::to_string(num_what_calls);
        return curr_what.c_str();
    }
    mutable int num_what_calls = 0;
    mutable std::string curr_what;
};

int main() {
    try {
        throw error_already_set();
    } catch (const std::runtime_error &e) {
        std::cout << e.what() << std::endl;
        std::cout << e.what() << std::endl;
    }
    return 0;
}

$ clang++ -std=c++11 error_already_set_up_cast.cpp 

$ a.out
1
2
$ 
```

Reply:
@rwgk You are right, nothing was wrong after all. The only qualm I have with this is the PGI having to be debug.